frontend/src/app/mentorship/programs/page.tsx

```diff
@@ -0,0 +1,168 @@
+
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  const debounceSearch = useCallback(
+    debounce((query: string) => {
+      setDebouncedQuery(query)
+    }, 500),
+    []
+  )
+
+  useEffect(() => {
+    debounceSearch(searchQuery)
+    return debounceSearch.cancel
+  }, [searchQuery, debounceSearch])
+
+  useEffect(() => {
+    const params = new URLSearchParams()
+    if (searchQuery) params.set('q', searchQuery)
+    if (page > 1) params.set('page', page.toString())
+
+    const queryString = params.toString()
+    const newUrl = queryString ? `?${queryString}` : window.location.pathname
+
+    if (window.location.search !== (queryString ? `?${queryString}` : '')) {
+      router.push(newUrl, { scroll: false })
+    }
+  }, [page, searchQuery, router])
+
+  const { data, loading, error } = useQuery(GET_PROGRAM_DATA, {
+    variables: {
+      page,
+      search: debouncedQuery,
+      mentorUsername: mentor,
+    },
+    fetchPolicy: 'cache-and-network',
+    errorPolicy: 'all',
+  })
+
+  const totalPages = data?.allPrograms?.totalPages
+  const currentPage = data?.allPrograms?.currentPage
+
+  useEffect(() => {
+    if (data?.allPrograms?.programs) {
+      setPrograms(data.allPrograms.programs)
+    }
+  }, [data])
+
+  useEffect(() => {
+    if (error) {
+      addToast({
+        description: 'Unable to fetch programs',
+        title: 'GraphQL Error',
+        timeout: 3000,
+        color: 'danger',
+        variant: 'solid',
+      })
+    }
+  }, [error])
+
+  const handleMyProgramsClick = () => {
+    const username = (session as SessionWithRole)?.user?.username
+    if (username) {
+      setMentor(username)
+      setPage(1)
+    }
+  }
+
+  const handleShowAllPrograms = () => {
+    setMentor('')
+    setPage(1)
+  }
+
+  const handleCreateButton = () => {
+    router.push('/mentorship/programs/create')
+  }
+
+  const renderProgramCard = (program: Program) => (
+    <Card
+      key={program.id}
+      title={program.name}
+      summary={program.description}
+      button={{
+        label: 'View Details',
+        icon: <FontAwesomeIconWrapper icon="fa-solid fa-right-to-bracket" />,
+        onclick: () => router.push(`/mentorship/programs/${program.key}`),
+      }}
+      timeline={{
+        start: program.startedAt,
+        end: program.endedAt,
+      }}
+      url={`/mentorship/programs/${program.key}`}
+    />
+  )
+
+  return (
+    <SearchPageLayout
+      isLoaded={!loading}
+      totalPages={totalPages}
+      currentPage={currentPage}
+      onPageChange={(p) => {
+        setPage(p)
+        window.scrollTo({ top: 0, behavior: 'auto' })
+      }}
+      onSearch={(q) => {
+        setSearchQuery(q)
+        setPage(1)
+      }}
+      searchQuery={searchQuery}
+      searchPlaceholder="Search Programs"
+      indexName="programs"
+    >
+      <div>
+        <div className="mt-8 flex justify-end gap-4">
+          {(session as SessionWithRole)?.user?.role === 'mentor' && (
```

Comment:
There's a faulty logic on how you handle sign in with GitHub. I wasn't able to see this div even after I updated my user to be a mentor. 
I then realized that when I signed in the role got overwritten for the `Nest user` model, setting it to `contributor` by default. I tested it several times, updating it back to `mentor` role and log out/log back in. 

Without this being fixed I'm not able to review this further.